client/allocrunner/taskrunner/device_hook.go

```diff
@@ -0,0 +1,95 @@
+package taskrunner
+
+import (
+	"context"
+	"fmt"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/client/devicemanager"
+	"github.com/hashicorp/nomad/plugins/device"
+	"github.com/hashicorp/nomad/plugins/drivers"
+)
+
+// deviceHook is used to retrieve device mounting information.
+type deviceHook struct {
+	logger log.Logger
+	dm     devicemanager.Manager
+}
+
+func newDeviceHook(dm devicemanager.Manager, logger log.Logger) *deviceHook {
+	h := &deviceHook{
+		dm: dm,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*deviceHook) Name() string {
+	return "devices"
+}
+
+func (h *deviceHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	if len(req.TaskResources.Devices) == 0 {
+		resp.Done = true
+		return nil
+	}
+
+	// Capture the responses
+	var reservations []*device.ContainerReservation
+	for _, req := range req.TaskResources.Devices {
+		// Ask the device manager for the reservation information
+		res, err := h.dm.Reserve(req)
+		if err != nil {
+			return fmt.Errorf("failed to reserve device %s: %v", req.ID(), err)
+		}
+
+		reservations = append(reservations, res)
+	}
+
+	// Build the response
+	for _, res := range reservations {
+		for k, v := range res.Envs {
+			if resp.Env == nil {
+				resp.Env = make(map[string]string)
+			}
+
+			resp.Env[k] = v
+		}
+
+		for _, m := range res.Mounts {
+			resp.Mounts = append(resp.Mounts, convertMount(m))
+		}
+
+		for _, d := range res.Devices {
+			resp.Devices = append(resp.Devices, convertDevice(d))
+		}
+	}
+
+	resp.Done = true
+	return nil
+}
+
+func convertMount(in *device.Mount) *drivers.MountConfig {
+	if in == nil {
+		return nil
```

Comment:
Do we reasonably expect nil values or is this defensive programming?  Also, I would suggest avoid inserting `nil`s into the `resp.Mounts` array if this is a reasonable case; and maybe logging it if it's unexpected.

Reply:
I can't find a reason these slices would contain nil items, so I'm removing these checks.